### PR TITLE
refactor: Use `node:` protocol for built-in Node.js modules

### PR DESCRIPTION
**Pull request by mdjermanovic**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Updates all modules to use `node:` protocol when requiring built-in Node.js modules.

#### What changes did you make? (Give an overview)

* Enabled `n/prefer-node-protocol` rule and let it autofix the whole project.
* Updated a few tests that use proxyquire.
* Updated webpack configuration to support `node:`.

#### Is there anything you'd like reviewers to focus on?

When testing whether the new `@eslint/config-array` package will work when used from eslint, I noticed that the webpack build (part of the browser test) fails with:

```js
ERROR in node:path
Module build failed: UnhandledSchemeError: Reading from "node:path" is not handled by plugins (Unhandled scheme).
Webpack supports "data:" and "file:" URIs by default.
You may need an additional plugin to handle "node:" URIs.
```

So, the first goal was to make the browser test (and eventually the Playground build on eslint.org) work. Then I switched to `node:` in a few places in eslint to additionally test the build, and in the end I decided to switch to `node:` in the entire project.

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -44,6 +46,12 @@ module.exports = {
         ]
     },
     plugins: [
+        new webpack.NormalModuleReplacementPlugin(
+            /^node:/u,
+            resource => {
+                resource.request = resource.request.replace(/^node:/u, "");
+            }
+        ),
```
Comment (mdjermanovic) [inline webpack.config.js:54]: This is a solution recommended in https://github.com/webpack/webpack/issues/13290

**Review by kecrily — COMMENTED** [review]
We can definitely use the [n/prefer-node-protocol rule](https://github.com/eslint-community/eslint-plugin-n/blob/master/docs/rules/prefer-node-protocol.md) at `eslint-config-eslint` to force the use of the node protocol

**Comment by mdjermanovic** [comment]
> We can definitely use the [n/prefer-node-protocol rule](https://github.com/eslint-community/eslint-plugin-n/blob/master/docs/rules/prefer-node-protocol.md) at `eslint-config-eslint` to force the use of the node protocol

Yeah, I enabled that rule in `eslint-config-eslint` in this PR.

https://github.com/eslint/eslint/pull/18434/files#diff-88a18842266c290a5fbf05c5d2829127c09c342d104e0dbf1ebc239a0a9e53b6

**Review by kecrily — APPROVED** [review]
Sorry for missing it. LGTM

**Review by fasttime — APPROVED** [review]
LGTM, thanks! Leaving open until tomorrow in case someone else wants to review.

**Review by aladdin-add — APPROVED** [review]

**Pull request merged**